File: day11_part2_2015.php

<?php

function readInput($filename) {
    $file = fopen($filename, "r");
    $input = fgets($file);
    fclose($file);
    return trim($input);
}

function findNextPassword($password) {
    while (true) {
        $password = incrementPassword($password);
        if (isValidPassword($password)) {
            break;
        }
    }
    return $password;
}

function incrementPassword($password) {
    $password = strrev($password);
    $carry = 1;
    for ($i = 0; $i < strlen($password); $i++) {
        $ascii = ord($password[$i]) + $carry;
        $password[$i] = chr($ascii > ord('z') ? ord('a') : $ascii);
        $carry = $ascii > ord('z') ? 1 : 0;
        if ($carry == 0) {
            break;
        }
    }
    return strrev($password);
}

function isValidPassword($password) {
    return hasStraight($password) && !containsInvalidLetters($password) && hasTwoPairs($password);
}

function hasStraight($password) {
    for ($i = 0; $i < strlen($password) - 2; $i++) {
        if (ord($password[$i]) + 1 == ord($password[$i + 1]) && ord($password[$i]) + 2 == ord($password[$i + 2])) {
            return true;
        }
    }
    return false;
}

function containsInvalidLetters($password) {
    return strpos($password, 'i') !== false || strpos($password, 'o') !== false || strpos($password, 'l') !== false;
}

function hasTwoPairs($password) {
    $count = 0;
    for ($i = 0; $i < strlen($password) - 1; $i++) {
        if ($password[$i] == $password[$i + 1]) {
            $count++;
            $i++;
        }
    }
    return $count >= 2;
}

$currentPassword = readInput("input.txt");
$firstNewPassword = findNextPassword($currentPassword);
$secondNewPassword = findNextPassword($firstNewPassword);
echo $secondNewPassword . PHP_EOL;

?>

File: objectivec/day11_part2_2015.m
#import <Foundation/Foundation.h>

NSString *readInput(NSString *filename) {
    NSError *error = nil;
    NSString *content = [NSString stringWithContentsOfFile:filename encoding:NSUTF8StringEncoding error:&error];
    return error ? nil : content;
}

NSString *incrementPassword(NSString *password) {
    NSMutableString *newPassword = [password mutableCopy];
    for (NSInteger i = newPassword.length - 1; i >= 0; i--) {
        unichar c = [newPassword characterAtIndex:i] + 1;
        if (c > 'z') {
            c = 'a';
        }
        [newPassword replaceCharactersInRange:NSMakeRange(i, 1) withString:[NSString stringWithFormat:@"%C", c]];
        if (c != 'a') break;
    }
    return newPassword;
}

BOOL hasStraight(NSString *password) {
    for (NSInteger i = 0; i < password.length - 2; i++) {
        if ([password characterAtIndex:i] + 1 == [password characterAtIndex:i + 1] &&
            [password characterAtIndex:i] + 2 == [password characterAtIndex:i + 2]) {
            return YES;
        }
    }
    return NO;
}

BOOL containsInvalidLetters(NSString *password) {
    NSCharacterSet *invalidSet = [NSCharacterSet characterSetWithCharactersInString:@"iol"];
    return [password rangeOfCharacterFromSet:invalidSet].location != NSNotFound;
}

BOOL hasTwoPairs(NSString *password) {
    NSInteger count = 0;
    for (NSInteger i = 0; i < password.length - 1; i++) {
        if ([password characterAtIndex:i] == [password characterAtIndex:i + 1]) {
            count++;
            i++; // Skip the next character
        }
    }
    return count >= 2;
}

BOOL isValidPassword(NSString *password) {
    return hasStraight(password) && !containsInvalidLetters(password) && hasTwoPairs(password);
}

NSString *findNextPassword(NSString *password) {
    while (YES) {
        password = incrementPassword(password);
        if (isValidPassword(password)) {
            break;
        }
    }
    return password;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *currentPassword = readInput(@"input.txt");
        NSString *firstNewPassword = findNextPassword(currentPassword);
        NSString *secondNewPassword = findNextPassword(firstNewPassword);
        NSLog(@"%@", secondNewPassword);
    }
    return 0;
}